gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx

```diff
@@ -0,0 +1,272 @@
+
+1. There are [five kinds of precision](https://github.com/python/mypy/blob/d807e097d142a88a48af93314d15ad87e41d2f19/mypy/stats.py#L27-L31). Full details are visible in the `--lineprecision-report`.
+2. Two kinds of precision, `EMPTY` and `UNANALYZED` convey no information, because there's nothing to analyze.
+3. A line is marked as precise, imprecise or any based on the expressions it uses.
+   - An expression that has type `Any` leads to precision `ANY`.
+   - I _think_ an expression that [involves `Any` but is not `Any`](https://github.com/python/mypy/blob/2907a4dea939e20ff39ef7d9ff85ae1a199a0ee8/mypy/stats.py#L422-L423) counts as imprecise. For instance, `Dict[str, Any]`.
+   - Remaining expressions have precision `PRECISE`.
+5. A line's precision is the worst of all its expressions' precisions.
+   - `ANY` is worse than `IMPRECISE`, which is worse than `PRECISE`.
+
+The "imprecision" number reported by mypy counts the number of lines [classified as `IMPRECISE` or `ANY`](https://github.com/python/mypy/blob/3bdef9fe6d401f3d2e0cacf4964bd315550c3394/mypy/xml/mypy-txt.xslt#L77-L78).
+
+On balance, my preferred metric is the line-level (im)precision percentage. There wasn't much difference between the two in my experience, but the colour-coded visualisation in the HTML report is a neat feature to have. Maybe in the future there could be a version of the HTML report that colour-codes each _expression_?
+
+
+## The larger typing ecosystem
+
+There are plenty of articles out there about the typing. As well as the [mypy blog itself](http://mypy-lang.blogspot.com/), see Daniele Varrazzo's [post on `psycopg3` (2020)](https://www.varrazzo.com/blog/2020/03/31/first-experience-mypy/), Dropbox's [blog post (2019)](https://dropbox.tech/application/our-journey-to-type-checking-4-million-lines-of-python), Zulip's [blog post (2016)](https://blog.zulip.com/2016/10/13/static-types-in-python-oh-mypy/), Glyph's [blog post on `Protocols` (2020)](https://glyph.twistedmatrix.com/2020/07/new-duck.html) and a follow-up [comparing them to `zope.interface` (2021)](https://glyph.twistedmatrix.com/2021/03/interfaces-and-protocols.html) and Nylas's [blog (2019)](https://www.nylas.com/blog/not-your-pie-mypy/). I'm sure there's plenty more out there.
+
+It's worth highlighting the `typeshed` project, which maintains stubs for the standard library, plus popular third-party libraries. I [submitted a PR](https://github.com/python/typeshed/pull/6150) to add a single type hint—it was a very pleasant experience! Microsoft has an [incubator](https://github.com/microsoft/python-type-stubs) of sorts for stubs too.
+
+### Other typecheckers
+
+After the sprint to improve coverage, I spent a short amount of time trying the alternative type checkers out there. This is by no means an authoritative, exhaustive survey—just my quick notes.
```

Comment:
Might be worth explicitly noting that "Mypy isn't the *only* typechecker in the Python ecosystem. Other companies have built their own tools to reason about type hints in Python, each with their own strengths and weaknesses."

Reply:
Thanks, d9f8bcc